device/iot-device-samples/send-event-x509/src/main/java/samples/com/microsoft/azure/sdk/iot/SSLContextBuilder.java

```diff
@@ -0,0 +1,246 @@
+/*
+ *  Copyright (c) Microsoft. All rights reserved.
+ *  Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+package samples.com.microsoft.azure.sdk.iot;
+
+import org.apache.commons.codec.binary.Base64;
+
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManagerFactory;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.*;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.UUID;
+
+/**
+ * Helper class that demonstrates how to build an SSLContext for x509 authentication from your public and private certificates,
+ * or how to build an SSLContext for SAS authentication from the default IoT Hub public certificates
```

Comment:
is the use-case here - devices that don't have the root certificates (digicert, baltimore etc) installed on their device, but need to trust those certs?

Reply:
For SAS based auth? Yes. Users of this SDK always have to have their SSLContext trust iothub's certs. This isn't something that the JVM just picks up from the physical device's certificate store

Reply:
For x509 auth, you would need to load the iothub public cert, your device's public cert, and your device's private key when building the SSLContext

Reply:
>Users of this SDK always have to have their SSLContext trust iothub's certs. This isn't something that the JVM just picks up from the physical device's certificate store

So this is something mandatory that needs to be passed by the users, and set by the SDK for every single TLS negotiation that the SDK will execute. In that case, why isn't SSLContext a mandatory argument?

Reply:
It's mandatory for establishing a connection to Iot hub, regardless of if it is x509 or SAS authentication. For SAS cases, the SDK can just build the SSLContext for the user which makes it optional at the constructor level

Reply:
ok, so for sas based authentication, what is the use-case for the customer wanting to pass this SSLContext to the SDK?

Reply:
Perhaps they have some gateway scenario set up and they need to trust a few more machines than just the iothub. Or perhaps they need to configure the TLS handshake to use a different version or something. By taking the SSLContext directly, we also let them do things like upgrade to TLS1.3 when the time comes without an SDK update

Reply:
Makes sense. I would add the note in comments for `ClientOptions.SSLConext` that the SDK will automatically create an SSL context to trust IoT hub certs, and if they pass in their custom SSLContext instance, they need to make sure to add IoT Hub's certificate info in there.

Reply:
So now a different question - the certs info that I see below in the sample, are these the public key info? How does the actual negotiation take place if the JVM cannot look up private key info in my device's certificate store/ I don't have the private key info for these certs?

Reply:
Good idea. I'll add a note to that effect in the javadoc for the clientOptions.sslContext

Reply:
The certs in the sample code's SSLContextBuilder class are indeed IoT Hub's public certificates. All SSLcontexts used must trust those certs to communicate with IoT Hub.

The user has to load their certs into the SSLContext from file or from string. Once that instance of SSLContext has the iothub public cert, their device's private key, and their device's public cert, then the SSL handshake will work for x509 auth.